autoload/maktaba/extension.vim

```diff
@@ -0,0 +1,159 @@
+""
+" @dict ExtensionRegistry
+" A registry for the extensions used a single plugin.  Extensions are
+" dictionaries with a plugin-specific interpretation
+"
+" Plugins should use |Plugin.GetExtensionRegistry| to gain access to their own
+" extension registry, and @function(maktaba#extension#GetRegistry) to gain
+" access to an extension registry for another plugin.
+
+
+" Extension registry objects, keyed by plugin name.
+if !exists('s:registries')
+  let s:registries = {}
+endif
+
+
+""
+" @private
+" Returns an extension registry for the given plugin name, creating it if
+" needed.
+" @throws WrongType if {plugin} is not a string.
+function! maktaba#extension#GetInternalRegistry(plugin) abort
+  call maktaba#ensure#IsString(a:plugin)
+
+  if has_key(s:registries, a:plugin)
+    return s:registries[a:plugin]
+  endif
+
+  " TODO: should this contain the plugin's name?
+  " TODO: is there a better way to construct a do-nothing _validator?
+  let l:registry = {
+      \ 'AddExtension': function('maktaba#extension#AddExtension'),
+      \ 'GetExtensions': function('maktaba#extension#GetExtensions'),
+      \ 'SetValidator': function('maktaba#extension#SetValidator'),
+      \ '_internal_extensions': [],
+      \ '_external_extensions': [],
+      \ '_validator': maktaba#function#FromExpr('0'),
+      \ '_external': 0,
+      \ }
+  let s:registries[a:plugin] = l:registry
+  return l:registry
+endfunction
+
+
+" Throws NotImplemented.  Used by the inaccessible functions on the external
+" interface.
+function! s:NotAccessible() abort
+  throw maktaba#error#NotImplemented('Not accessible by external plugins.')
```

Comment:
I'd almost say to just inline this into the 2 places you call it. Certainly doesn't save any typing this way. =)